From 050ef4fab0a4679b1435d13849bf229922b9637b Mon Sep 17 00:00:00 2001
From: Cao Zou <cao.zou@windriver.com>
Date: Mon, 3 Jul 2017 09:54:47 +0800
Subject: [PATCH] RT: sh: add a kgdb check in console write

Add a kgdb check in tty console_write function, it will fix a call trace
as follow:
[<ffff80000008abf0>] dump_backtrace+0x0/0x128
[<ffff80000008ad3c>] show_stack+0x24/0x30
[<ffff80000096c6b0>] dump_stack+0xac/0xe8
[<ffff8000000d04b8>] ___might_sleep+0x248/0x260
[<ffff800000973150>] rt_spin_lock+0x30/0x68
[<ffff80000059e1fc>] serial_console_write+0x17c/0x180
[<ffff800000162568>] vkdb_printf+0x1c0/0x920_

Signed-off-by: Cao Zou <cao.zou@windriver.com>
---
 drivers/tty/serial/sh-sci.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/drivers/tty/serial/sh-sci.c b/drivers/tty/serial/sh-sci.c
index c9e5c02..77e0a9d 100644
--- a/drivers/tty/serial/sh-sci.c
+++ b/drivers/tty/serial/sh-sci.c
@@ -44,6 +44,7 @@
 #include <linux/pm_runtime.h>
 #include <linux/scatterlist.h>
 #include <linux/serial.h>
+#include <linux/kdb.h>
 #include <linux/serial_sci.h>
 #include <linux/sh_dma.h>
 #include <linux/slab.h>
@@ -3067,7 +3068,7 @@ static void serial_console_write(struct console *co, const char *s,
 		locked = 0;
 	else
 #endif
-	if (oops_in_progress)
+	if (oops_in_progress || in_kdb_printk())
 		locked = spin_trylock_irqsave(&port->lock, flags);
 	else
 		spin_lock_irqsave(&port->lock, flags);
-- 
1.7.5.4

